feat(cli): Add --allowed-tools flag to bypass tool confirmation (#2417)

werdnum opened the pull request:
This PR introduces the `--allowed-tools` flag and a corresponding `allowedTools` setting, allowing users to specify tools that can run without confirmation. This is useful for trusted and frequently used commands like `git status`.

I have chosen `--allowed-tools` (instead of `--auto-approved-tools` as I had initially) because at the moment its functionality is quite complementary to `coreTools` (if you are in YOLO/unattended mode, `coreTools` is useful and `allowedTools` is not, and if you're in interactive mode, `coreTools` is not that useful but `allowedTools` is), so we may want to merge `coreTools` into `allowedTools` – and `allowedTools` works well for both use cases.

Addresses #2417

**Key Changes:**

*   **New Feature:**
    *   Adds the `--allowed-tools` command-line flag.
    *   Adds the `allowedTools` setting to the configuration schema.
    *   Updates documentation in `docs/cli/configuration.md`.

*   **Refactoring & Bug Fix:**
    *   To support this feature correctly for shell commands with arguments, the tool filtering logic was refactored.
    *   The core matching logic is now centralized in the `doesToolInvocationMatch` function located in `packages/core/src/utils/shell-utils.ts`.
    *   This refactoring unifies how `coreTools`, `excludedTools`, and the new `allowedTools` are handled, resolving a circular dependency.
    *   The redundant `packages/core/src/utils/tool-utils.ts` file has been removed.
    *   Fixes a bug where matching shell command patterns with arguments (e.g., `run_shell_command(git status)`) was not working correctly.

werdnum commented:
/gemini review

abhipatel12 commented on packages/core/src/core/coreToolScheduler.test.ts:
```diff
@@ -53,6 +53,7 @@ describe('CoreToolScheduler', () => {
       getUsageStatisticsEnabled: () => true,
       getDebugMode: () => false,
       getApprovalMode: () => ApprovalMode.DEFAULT,
+      getAutoApprovedTools: () => [],
```
Can we add a test for the positive flow here? So something where the test would require approval, but b/c we added it to the approved list, it is bypassed?

abhipatel12 reviewed (commented):
Hey Andrew thanks for this PR!

Left a few comments, let me know what you think. Some naming questions about the flag + some relocation just in case you had thoughts on that before you made any changes!

werdnum commented on packages/core/src/utils/shell-utils.ts:
```diff
@@ -4,10 +4,79 @@
+ *     invocations where the arguments start with that prefix.
+ * @returns True if the invocation matches any pattern, false otherwise.
+ */
+export function doesToolInvocationMatch(
```
Heh, Gemini initially put it in tool-utils and I had it merge into shell-utils for simplicity, but I don't feel strongly. I'll change it back.

werdnum commented:
Rebase was a bit of a pain.

Of note, I adjusted the order of some of the permissions checks (in particular we always check if a tool requires confirmation), this made one test fail because it was looking for that method to never be called. IMO that is kind of overspecified so I just deleted the assertion.

werdnum commented:
/gemini review

allenhutchison commented on packages/cli/src/config/settingsSchema.ts:
```diff
@@ -314,6 +314,16 @@ export const SETTINGS_SCHEMA = {
     description: 'Paths to core tool definitions.',
     showInDialog: false,
   },
+  autoApprovedTools: {
```
I don't have a strong opinion here, but personally like the verbosity of auto-approve-tools since it does exactly what it says. Failing that then I'd go with allowedTools to try to stay consistent with norms set in the industry.
